extensions/ql-vscode/src/eval-log-tree-builder.ts

```diff
@@ -0,0 +1,52 @@
+import { ChildEvalLogTreeItem, EvalLogTreeItem } from './eval-log-visualizer';
+import { EvalLogData as EvalLogData } from './pure/log-summary-parser';
+
+export default class EvalLogTreeBuilder {
+    private evalLogDataItems: EvalLogData[];
+    
+    constructor(evaluatorLogDataItems: EvalLogData[]) {
+        this.evalLogDataItems = evaluatorLogDataItems;
+    }
+
+    async getRoots(): Promise<EvalLogTreeItem[]> {
+        return await this.parseRoots();
+    }
+
+    private async parseRoots(): Promise<EvalLogTreeItem[]> {
+        const roots: EvalLogTreeItem[] = [];
+
+        // For each item, create a TreeItem object with appropriate parents/children 
+        this.evalLogDataItems.forEach(logDataItem => {
+
+            // TODO(angelapwen): Once the viewer can show logs for multiple queries, then queries should be the root level.
+            // Every object is a root except the RA for now
+            const label = `${logDataItem.predicateName} - ${logDataItem.resultSize} tuples in ${logDataItem.millis} ms for query ${logDataItem.queryCausingWork}`;
+            const rootItem: EvalLogTreeItem = {
+                label: label,
+                children: [] // Will assign pipeline items as children shortly.
+            };
+
+            for (const [pipelineName, steps] of Object.entries(logDataItem.ra)) {
+                const pipelineLabel = `Pipeline ID: ${pipelineName}`;
+                const pipelineItem: ChildEvalLogTreeItem = {
+                    label: pipelineLabel,
+                    parent: rootItem,
+                    children: [] // Will assign step items as children shortly.
+                };
+                rootItem.children.push(pipelineItem);
+        
+                steps.forEach( (step: string) => {
+                    const stepLabel = step;
+                    const stepItem: ChildEvalLogTreeItem =  {
+                        label: stepLabel,
+                        parent: pipelineItem,
+                        children: []
+                    };
+                    pipelineItem.children.push(stepItem);
+                });
```

Comment:
Oops, I committed this but reverted because of a type issue. `steps.map(...)` returns a `void[]` but it needs to be a `ChildEvalLogTreeItem[]`. I tried a few different type coercions/ways to get each item in the array to be of type `ChildEvalLogTreeItem` but I'm not sure if I'm just not finding the right syntax or if this isn't possible 😄

Reply:
Ugh my fault for not trying it out myself. Put an extra pair of parentheses around the `{}` object, so you get
```js
steps.map((step: string) => ({
  label: step,
  parent: pipelineItem,
  children: []
}));
```

Without the extra parentheses, I think the opening `{` is treated as starting the body of a function (which has no return statement, hence the `void` type), instead of starting an object literal like we intended.

(The `: string` type annotation is optional btw; TS will infer it.)

Reply:
Ah, I see! I had thought that it was an object of type `void` because TypeScript couldn't figure out how to make it of type `ChildEvalLogTreeItem[]` but it was actually inferring that it was a function returning `void` 🤯 

have made the change now.